File: py2_energy_test_final_feasible/shortest-matching-substring/shortest-matching-substring_benchmark.py
# coding: utf-8
import time

# ========== PASSING PYTHON 2 SOLUTION (100% UNMODIFIED) ==========
# Time:  O(n + m)
# Space: O(n + m)

# kmp, two pointers (three pointers)
class Solution(object):
    def shortestMatchingSubstring(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: int
        """
        INF = float("inf")
        def getPrefix(pattern):
            prefix = [-1]*len(pattern)
            j = -1
            for i in xrange(1, len(pattern)):
                while j+1 > 0 and pattern[j+1] != pattern[i]:
                    j = prefix[j]
                if pattern[j+1] == pattern[i]:
                    j += 1
                prefix[i] = j
            return prefix

        def KMP(text, pattern):
            if not pattern:
                for i in xrange(len(text)+1):
                    yield i
                return
            prefix = getPrefix(pattern)
            j = -1
            for i in xrange(len(text)):
                while j+1 > 0 and pattern[j+1] != text[i]:
                    j = prefix[j]
                if pattern[j+1] == text[i]:
                    j += 1
                if j+1 == len(pattern):
                    yield i-j
                    j = prefix[j]
        
        parts = p.split('*')
        a = parts[0] if len(parts) > 0 else ''
        b = parts[1] if len(parts) > 1 else ''
        c = parts[2] if len(parts) > 2 else ''
        n = len(s)
        la, lb, lc = len(a), len(b), len(c)
        result = INF
        j = k = 0
        jt = KMP(s, b)
        kt = KMP(s, c)
        for i in KMP(s, a):
            while j != -1 and j < i+la:
                j = next(jt, -1)
            if j == -1:
                break
            while k != -1 and k < j+lb:
                k = next(kt, -1)
            if k == -1:
                break
            result = min(result, (k+lc)-i)
        return result if result != INF else -1 
            

# Time:  O(n + m)
# Space: O(n + m)
# kmp, two pointers (three pointers)
class Solution2(object):
    def shortestMatchingSubstring(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: int
        """
        INF = float("inf")
        def getPrefix(pattern):
            prefix = [-1]*len(pattern)
            j = -1
            for i in xrange(1, len(pattern)):
                while j+1 > 0 and pattern[j+1] != pattern[i]:
                    j = prefix[j]
                if pattern[j+1] == pattern[i]:
                    j += 1
                prefix[i] = j
            return prefix

        parts = p.split('*')
        a = parts[0] if len(parts) > 0 else ''
        b = parts[1] if len(parts) > 1 else ''
        c = parts[2] if len(parts) > 2 else ''
        n = len(s)
        la, lb, lc = len(a), len(b), len(c)
        prefix1 = getPrefix(a+'#'+s)
        prefix2 = getPrefix(b+'#'+s)
        prefix3 = getPrefix(c+'#'+s)
        result = INF
        i = j = k = 0
        while i+lb+lc < n:
            while i < n and prefix1[la+1+i]+1 != la:
                i += 1
            if i == n:
                break
            while j < n and not (j >= i+lb and prefix2[lb+1+j]+1 == lb):
                j += 1
            if j == n:
                break
            while k < n and not (k >= j+lc and prefix3[lc+1+k]+1 == lc):
                k += 1
            if k == n:
                break
            result = min(result, k-(i-la))
            i += 1
        return result if result != INF else -1

# ==================================================================

# ========== TEST HARNESS ==========
def main():
    iterations = 1000

    # Test cases extracted from C++ (convert C++ syntax to Python 2)
    test_cases = [
        ("abc", "a*b*c"),            # straightforward match
        ("abbbc", "ab*b*c"),         # multiple occurrences of middle part
        ("xbxcxd", "*b*c"),          # empty a
        ("abccba", "ab**ba"),        # empty middle part
        ("aaaaa", "aa*aa*aa"),       # impossible due to non-overlap constraint
        ("zz", "**"),                # all parts empty -> shortest 0
        ("abcde", "x*y*z"),          # no match at all
        ("xyzaaxyzb", "xy*aa*zb"),   # spaced parts
        ("cababcab", "ca*ba*b"),     # multiple choices for minimal window
        ("helloworld", "world"),     # no star in pattern
    ]

    # Warm-up
    sol = Solution2()
    for test in test_cases:
        if isinstance(test, tuple):
            sol.shortestMatchingSubstring(*test)
        else:
            sol.shortestMatchingSubstring(test)

    # Benchmark
    results = []
    start_time = time.time()

    for _ in xrange(iterations):
        sol = Solution2()
        for test in test_cases:
            if isinstance(test, tuple):
                result = sol.shortestMatchingSubstring(*test)
            else:
                result = sol.shortestMatchingSubstring(test)
            results.append(result)

    elapsed_time = time.time() - start_time
    checksum = sum(hash(str(r)) for r in results) % (10**9 + 7)
    print("Checksum: {}, Total time: {:.6f}s".format(checksum, elapsed_time))

if __name__ == "__main__":
    main()

File: transpiled_after/shortest-matching-substring/cpp/shortest-matching-substring_benchmark.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <climits>
#include <chrono>

using namespace std;
using namespace std::chrono;

// Time:  O(n + m)
// Space: O(n + m)
// kmp, two pointers (three pointers)
class Solution {
public:
    int shortestMatchingSubstring(string s, string p) {
        auto splitByStar = [](const string& str) -> vector<string> {
            vector<string> parts;
            string cur;
            for (char ch : str) {
                if (ch == '*') {
                    parts.push_back(cur);
                    cur.clear();
                } else {
                    cur.push_back(ch);
                }
            }
            parts.push_back(cur);
            return parts;
        };

        auto getPrefix = [&](const string& pattern) -> vector<int> {
            int n = static_cast<int>(pattern.size());
            vector<int> prefix(n, -1);
            int j = -1;
            for (int i = 1; i < n; ++i) {
                while (j + 1 > 0 && pattern[j + 1] != pattern[i]) {
                    j = prefix[j];
                }
                if (pattern[j + 1] == pattern[i]) {
                    ++j;
                }
                prefix[i] = j;
            }
            return prefix;
        };

        auto KMP = [&](const string& text, const string& pattern) -> vector<int> {
            vector<int> res;
            if (pattern.empty()) {
                for (int i = 0; i <= static_cast<int>(text.size()); ++i) {
                    res.push_back(i);
                }
                return res;
            }
            vector<int> prefix = getPrefix(pattern);
            int j = -1;
            for (int i = 0; i < static_cast<int>(text.size()); ++i) {
                while (j + 1 > 0 && pattern[j + 1] != text[i]) {
                    j = prefix[j];
                }
                if (pattern[j + 1] == text[i]) {
                    ++j;
                }
                if (j + 1 == static_cast<int>(pattern.size())) {
                    res.push_back(i - j);
                    j = prefix[j];
                }
            }
            return res;
        };

        vector<string> parts = splitByStar(p);
        // Expect exactly three parts: a, b, c
        string a = parts.size() > 0 ? parts[0] : string();
        string b = parts.size() > 1 ? parts[1] : string();
        string c = parts.size() > 2 ? parts[2] : string();

        int n = static_cast<int>(s.size());
        int la = static_cast<int>(a.size());
        int lb = static_cast<int>(b.size());
        int lc = static_cast<int>(c.size());

        vector<int> occA = KMP(s, a);
        vector<int> occB = KMP(s, b);
        vector<int> occC = KMP(s, c);

        int result = INT_MAX;
        int idxB = 0, idxC = 0;

        for (int iIdx = 0; iIdx < static_cast<int>(occA.size()); ++iIdx) {
            int i = occA[iIdx];
            while (idxB < static_cast<int>(occB.size()) && occB[idxB] < i + la) {
                ++idxB;
            }
            if (idxB == static_cast<int>(occB.size())) {
                break;
            }
            while (idxC < static_cast<int>(occC.size()) && occC[idxC] < occB[idxB] + lb) {
                ++idxC;
            }
            if (idxC == static_cast<int>(occC.size())) {
                break;
            }
            result = min(result, (occC[idxC] + lc) - i);
        }

        return result == INT_MAX ? -1 : result;
    }
};

int main() {
    // Define 10 diverse test inputs (s and p lists)
    vector<string> s_tests = {
        "abc",           // straightforward match
        "abbbc",         // multiple occurrences of middle part
        "xbxcxd",        // empty a
        "abccba",        // empty middle part
        "aaaaa",         // impossible due to non-overlap constraint
        "zz",            // all parts empty -> shortest 0
        "abcde",         // no match at all
        "xyzaaxyzb",     // spaced parts
        "cababcab",      // multiple choices for minimal window
        "helloworld"     // no star in pattern
    };
    vector<string> p_tests = {
        "a*b*c",
        "ab*b*c",
        "*b*c",
        "ab**ba",
        "aa*aa*aa",
        "**",
        "x*y*z",
        "xy*aa*zb",
        "ca*ba*b",
        "world"
    };

    Solution sol;
    const int iterations = 1000;

    volatile long long checksum = 0; // volatile to discourage optimization
    auto start = high_resolution_clock::now();
    for (int iter = 0; iter < iterations; ++iter) {
        for (size_t ti = 0; ti < s_tests.size(); ++ti) {
            int res = sol.shortestMatchingSubstring(s_tests[ti], p_tests[ti]);
            checksum += res;
        }
    }
    auto end = high_resolution_clock::now();
    auto duration_us = duration_cast<microseconds>(end - start).count();

    cout << "Checksum: " << checksum << "\n";
    cout << "Time (microseconds): " << duration_us << "\n";
    return 0;
}